### PR TITLE
Enable `dim=None` for `torch.mean`

### LINKED ISSUE #29137: torch.sum(tensor, dim=()) is different from np.sum(arr, axis=())
## 🐛 Bug

`torch.sum(tensor, dim=())` performs a full reduce, while `np.sum(arr, axis=())` performs no reduce.

## To Reproduce

```
import torch
import numpy as np
arr = np.array([1, 2, 3])
tensor = torch.from_numpy(arr)

tensor.sum(())  # gives tensor(6)
arr.sum(())  # gives np.array([1, 2, 3])
```
## Expected behavior

These should probably be the same. Also, I don't think it makes sense for `tensor.sum(dim=())` to do a full reduce. This special casing caused the bug at https://github.com/pytorch/pytorch/issues/28993 and I suspect the gradient for things like torch.var may be incorrect right now.

## Environment

Pytorch master

## Additional context

Some people who have been involved in multidim support who may be interested: @t-vi, @SsnL, @umanwizard @gchanan. Please let me know your opinions on this subject.

cc @ezyang @gchanan @zou3519 @jerryzh168 @SsnL

### PR DESCRIPTION
**Pull request by kurtamohler**
Part of #79525

This will require coordination with XLA before merging, just like #79881

### CODE REVIEW

**Review by kulinseth — COMMENTED** [review]
MPS side changes look good.

**Review by albanD — APPROVED** [review]
Change SGTM
Is the xla side ready?

**Comment by kurtamohler** [comment]
Hey @JackCaoG, this PR breaks a function signature in XLA, like #79881 did. I will be submitting a steady stream of PRs like this over the near future

**Comment by JackCaoG** [comment]
@wonjoolee95 Can you help drafting a XLA pr similar to https://github.com/pytorch/xla/pull/3554?

**Comment by wonjoolee95** [comment]
We have opened https://github.com/pytorch/xla/pull/3752 on XLA's side. Build looking good locally, we'll wait for a CI (which usually takes an hour or so) then we'll merge it. I'll post updates here.

**Comment by wonjoolee95** [comment]
@kurtamohler, the XLA's PR is green so far but CPU/GPU tests are still running. Let's coordinate a merge on Monday, if that works with you.

**Comment by kurtamohler** [comment]
Sounds good, thanks @wonjoolee95

**Comment by wonjoolee95** [comment]
Hi @kurtamohler, are we good to merge this today? If so, I'll go ahead and merge the pt/xla PR now. Thanks!

**Comment by kurtamohler** [comment]
@wonjoolee95, sure thing! I'll update this PR as soon as yours is merged

**Comment by wonjoolee95** [comment]
Thanks, @kurtamohler! Just merged the PR, the new commit hash for XLA is `c88a86b044542c278099f32f58e286415681c600`.

**Comment by kurtamohler** [comment]
> Merge failed due to Refusing to merge as mandatory check(s) [pull](https://github.com/pytorch/pytorch/commit/7987aff0ba4c3f05bf6aa40878dcbefa0598c0a2/checks?check_suite_id=7512313283) failed for rule superuser Raised by https://github.com/pytorch/pytorch/actions/runs/2734329189

I don't understand what this means

**Comment by albanD** [comment]
It refused to merge because there is a test that is red in the CI.
Namely: https://github.com/pytorch/pytorch/runs/7505998993?check_suite_focus=true

**Comment by kurtamohler** [comment]
Oh I see. The `android-emulator-build-test` and `linux-vulkan-bionic-py3.7-clang9` failures happened before, and I mistakenly thought they were flaky/upstream. I'll look into them. They might be tricky, the error messages don't give much info

**Comment by JackCaoG** [comment]
@kurtamohler should I revert xla pr?

**Comment by kurtamohler** [comment]
@JackCaoG, yeah sorry, it will probably take me a while to figure out the issue

**Comment by JackCaoG** [comment]
no worry, I think we can start experiment the new landing process I proposed in https://github.com/pytorch/pytorch/issues/82033 next time. In that case landing won't break either of the CI

**Comment by kurtamohler** [comment]
@wonjoolee95 and @JackCaoG , all tests are passing now (except xla of course). We can try merging again when you're ready. Feel free to ping me on the PyTorch slack if it's easier

**Comment by wonjoolee95** [comment]
Hi @kurtamohler, thanks for taking care of this! Let's coordinate a merge tomorrow, I'll message you through PyTorch slack tomorrow morning.

**Comment by JackCaoG** [comment]
Let's give https://github.com/pytorch/pytorch/issues/82033 a try tmr 😄

**Comment by wonjoolee95** [comment]
Hmm, xla CI build seems to be failing while building tests due to import errors:
```
[ 95%] Building CXX object CMakeFiles/test_ptxla.dir/torch_xla_test.cpp.o
In file included from /var/lib/jenkins/workspace/xla/test/cpp/cpp_test_util.cpp:14:
In file included from /var/lib/jenkins/workspace/xla/torch_xla/csrc/torch_util.h:8:
/opt/conda/lib/python3.7/site-packages/torch/include/torch/csrc/lazy/core/dynamic_ir.h:19:10: fatal error: 'torch/csrc/lazy/ts_backend/ts_node.h' file not found
#include <torch/csrc/lazy/ts_backend/ts_node.h>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /var/lib/jenkins/workspace/xla/test/cpp/test_aten_xla_tensor.cpp:12:
In file included from /var/lib/jenkins/workspace/xla/torch_xla/csrc/torch_util.h:8:
/opt/conda/lib/python3.7/site-packages/torch/include/torch/csrc/lazy/core/dynamic_ir.h:19:10: fatal error: 'torch/csrc/lazy/ts_backend/ts_node.h' file not found
#include <torch/csrc/lazy/ts_backend/ts_node.h>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by JackCaoG** [comment]
yeounoh had a fix https://github.com/pytorch/pytorch/pull/82384

**Comment by kurtamohler** [comment]
Looks like this merge caused a failure on master: https://github.com/pytorch/pytorch/runs/7570327286?check_suite_focus=true

The CI hud shows it failing for this commit and all the ones after it: https://ezyang.github.io/pytorch-ci-hud/build2/pytorch-master

**Comment by wonjoolee95** [comment]
I've looked a couple of CI failures on the last few PyTorch commits, the failures actually seems to be due to:
```
var/lib/jenkins/workspace/xla/torch_xla/csrc/torch_util.cpp:12:26: error: no member named 'SymbolicIntNode' in namespace 'c10'
    std::shared_ptr<c10::SymbolicIntNode> symbolicIntNode =
                    ~~~~~^
/var/lib/jenkins/workspace/xla/torch_xla/csrc/torch_util.cpp:12:43: error: C++ requires a type specifier for all declarations
    std::shared_ptr<c10::SymbolicIntNode> symbolicIntNode =
                                          ^
/var/lib/jenkins/workspace/xla/torch_xla/csrc/torch_util.cpp:13:14: error: no member named 'toSymbolicIntNode' in 'c10::SymInt'
        size.toSymbolicIntNode();
        ~~~~ ^
/var/lib/jenkins/workspace/xla/torch_xla/csrc/torch_util.cpp:15:48: error: no member named 'SymbolicIntNode' in namespace 'torch::lazy'
        std::dynamic_pointer_cast<torch::lazy::SymbolicIntNode>(
                                  ~~~~~~~~~~~~~^
/var/lib/jenkins/workspace/xla/torch_xla/csrc/torch_util.cpp:26:54: warning: moving a temporary object prevents copy elision [-Wpessimizing-move]
    auto size_node = torch::lazy::MakeNode<Constant>(std::move(
                                                     ^
/var/lib/jenkins/workspace/xla/torch_xla/csrc/torch_util.cpp:26:54: note: remove std::move call here
    auto size_node = torch::lazy::MakeNode<Constant>(std::move(
                                                     ^~~~~~~~~~
```
Which is addressed by an XLA PR that was merged earlier this afternoon https://github.com/pytorch/xla/pull/3793. 

This seems to be due to this PR's new XLA commit pin not including the https://github.com/pytorch/xla/pull/3793. It's a mistake on the merge process, my XLA companion PR (https://github.com/pytorch/xla/pull/3782) should have rebased on master to pull in https://github.com/pytorch/xla/pull/3782 and have this PR pinned on that new commit id (although this is a bit of an edge case since there were multiple breaking changes happening at the same time). 

I believe the nightly PyTorch bot updates the XLA commit id and should fix this build CI failures, since the current XLA head is green and include all the needed changes (although I'm not too familiar with exactly when/how PyTorch bot updates the XLA commit id).

cc @JackCaoG

**Comment by JackCaoG** [comment]
I think xla hash has been updated to latest according to https://github.com/pytorch/pytorch/blob/master/.github/ci_commit_pins/xla.txt, we should be fine.
